pubsub.py

```diff
@@ -0,0 +1,65 @@
+# Copyright 2020 Google, LLC.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module to handle input from Google Monitoring.
+
+This module defines functions and errors to handle input from Google Monitoring,
+such as Pub/Sub notifications.
+"""
+
+import base64
+import binascii
+
+
+class DataParseError(Exception):
```

Comment:
Sounds good. I made the base Error class in philips_hue since there were multiple user-defined errors but I see how this is a good practice regardless :)

Reply:
Just to clarify: the idiom is to provide a module-level base Exception called `Error`, so that callers can say: `handle all of the errors originating from pubsub.py`. Python modules correspond to files: https://docs.python.org/3/tutorial/modules.html

So under this idiom, you would define an Error(Exception) per .py file.

The problem with defining a single base Error class in philips_hue is that it leads to code like:

```
try:
    pubsub.some_function()
except philips_hue.Error:
    # handle error
```

Which doesn't really make sense. If you define an Error per module, it permits code like:

```
try:
    pubsub.some_function()
except pubsub.Error:
    # handle all pubsub related errors

try:
   philips_hue.some_function()
except philips_hue.Error:
   # handle all philips hue errors
```

Note, that there is another common idiom. Which is to define a directory or project wide errors.py. E.g.

```
# errors.py

class Error(Exception):

class FooError(Error):

class BarError(Error):

# file1.py

try:
    # some code
except errors.FooError
    # handle

# file2.py

try:
   # some code
except errors.Error:
   # very broad exception, handle any error that can be raised in this directory or project.
```

This allows you to share errors across modules. I just mention this. I don't think that this is especially useful for the pubsub error handling. I think a pubsub module level error makes sense.

Reply:
Got it, thanks for the clarification! Sorry I think I didn't explain myself clearly when I mentioned adding the Error class to philips_hue; I knew it just applied to the philips_hue module, I just meant that because I defined multiple errors in philips_hue, I found that adding a base error class there was a good idea but thought that since I only had one error type in pubsub it was fine to just inherit from Exception. The idiom makes sense regardless of the number of exceptions defined though so thanks for bringing it up!